crates/ruff_python_semantic/src/model.rs

```diff
@@ -1530,14 +1530,28 @@ impl<'a> SemanticModel<'a> {
             return true;
         }
 
-        false
+        // Ex) `Union[Union[int, int]]` or `Union[int | int]`
+        self.current_expression_parent()
+            .and_then(Expr::as_subscript_expr)
+            .is_some_and(|parent| self.match_typing_expr(&parent.value, "Union"))
```

Comment:
I don't understand why we have to consider `current_expression_parent()` *and* `current_expression_grandparent()` for subscript-based unions, but we only have to consider `current_expression_parent()` for `|`-based unions. Does that mean that this function will recognise this as being a nested union:

```py
Union[Union[int, str]]
```

but not necessarily this?

```py
Union[int, str] | Union[bytes, list]
```

Reply:
I think this is because the PEP604-style unions are always binary operations, but I need to double check.

Reply:
Okay, I think I now understand why we must consider both the `current_expression_grandparent()` and the `current_expression_parent()` for subscript unions. Please tell me if I misunderstand anything here.

It's because there's a difference in AST structure for these two nested unions:

```py
A = Union[Union[int]]
B = Union[Union[int], bytes]
```

For union `A`, there, the AST structure looks like this (simplifying a little; this isn't valid Rust, obviously):

```rs
Subscript {
    value: Name("Union")
    slice: Subscript {
        value: Name("Union")
        slice: Name("int")
    }
}
```

But for union `B`, the AST structure looks like this:

```rs
Subscript {
    value: Name("Union")
    slice: Tuple([
        Subscript {
            value: Name("Union")
            slice: Name("int")
        },
        Name("bytes")
    ])
}
```

So from the perspective of the inner union for `A`, the outer union is the `current_expression_parent` (there's no intermediate `Tuple` expression because there's only one element in the subscript slice). But from the perpespective of the inner union for `B`, the outer union is the `current_expression_grandparent` (there's an intermediate `Tuple` expression because there's more than one element in the subscript slice).

Is that correct? If so, please could you expand on the comments you have here, as it took me quite a while to understand all this :-)

I also think you could make the implementation more efficient and accurate (accurate in that `current_expression_parent` must be a `Tuple` as well as `current_expression_grandparent` being a `Union` subscript in order for it to be a nested union). Something like this?

```rs
    /// Return `true` if the model is in a nested union expression (e.g., the inner `Union` in
    /// `Union[Union[int, str], float]`).
    pub fn in_nested_union(&self) -> bool {
        let mut parent_expressions = self.current_expressions().skip(1);

        match parent_expressions.next() {
            Some(Expr::Subscript(parent)) => self.match_typing_expr(&parent.value, "Union"),
            Some(Expr::Tuple(_)) => parent_expressions
                .next()
                .and_then(Expr::as_subscript_expr)
                .is_some_and(|grandparent| self.match_typing_expr(&grandparent.value, "Union")),
            Some(Expr::BinOp(bin_op)) => bin_op.op.is_bit_or(),
            _ => false,
        }
```

Reply:
100%, this is much cleaner 👍 I'll expand this with comments and rewrite `in_nested_literal` in a similar fashion.